keps/sig-node/0030-qos-class-cpu-affinity.md

```diff
@@ -0,0 +1,152 @@
+
+## Motivation
+
+Related issues: https://github.com/kubernetes/kubernetes/issues/67355
+
+This behavior is documented in the [Intel optimization guides](https://computing.llnl.gov/tutorials/linux_clusters/intelAVXperformanceWhitePaper.pdf) as well as being [well described](https://www.anandtech.com/show/11544/intel-skylake-ep-vs-amd-epyc-7000-cpu-battle-of-the-decade/8) in consumer computing publications. 
+
+### Goals
+ - Allow operators to perform some basic system partitioning to address the scenarios outlined above.
+
+### Non-Goals
+ - TBD
+
+## Proposal
+
+Main idea: Allow operators to hint container CPU affinity based on Quality-of-Service class.
+
+### Changes to existing components
+
+Add two new configuration values to the Kubelet:
+ - `be-cpuset` (CPUs on which BestEffort can run)
+ - `other-cpuset` (CPUs on which Burstable and Guaranteed can run)
+
+Both `be-cpuset` and `other-cpuset` default to all available CPUs.
+
+Update the Kubelet's computation for the allocatable quantity of CPUs:
+
+`min((ALL – sys.reserved - kube.reserved), #OTHER)`
+
+Enforce configured CPU affinity by updating the CPU manager.
+
+Update the CPU Manager static policy to initialize the allocatable CPU pool from the configured `OTHER` cpuset.
+
+### Examples
+
+The following illustrates how the motivating use cases can be satisfied given the proposed changes. Let `ALL`, `ISOLCPUS`, `BE` and `OTHER` be sets of CPU IDs.
+
+#### Variant 1: Batch workloads targeting AVX, run as Best-Effort
+
+Configure `BE` to be a subset of `ALL`. The size of the chosen set represents an acceptable trade-off for your environment, effectively limiting the frequency downscaling due to SIMD and TDP (by capping the number of physical cores issuing AVX instructions.)
+
+![qos-be-cpuset](./0030-qos-be-cpuset.png)
+
+*Figure 1: Example BestEffort cpuset. If AVX instructions are only run as `BE`, then this caps the number of physical cores that can issue AVX instructions to one.*
+
+**Optional extension (out of tree):**
+
+An operator might prefer local users to be explicit about mapping pods that use a particular ISA extension to a target QoS class. A validating admission controller could validate for example that pods labeled "AVX" run in the BestEffort QoS class. A more opinionated admission controller could mutate the pod spec, removing resource requests from "AVX" pods, thereby forcing them into the BestEffort class. This or similar mechanisms could exist outside of core Kubernetes, following recommendations from hardware vendors.
+
+#### Variant 2: Workloads targeting AVX, run as Burstable or Guaranteed
+
+Enable the existing CPU manager static policy. Use existing ability to advertise extended resources to allow a limited number of "AVX cores". Enforce that if a container consumes the "AVX cores" resource:
+- AVX Cores = CPU
+- QoS class is Guaranteed
```

Comment:
What guarantees that a Gu pods consuming Avx will not disrupt another Avx pod not consuming Avx?

Reply:
They still might, but at least the AVX and non-AVX pods wouldn't run on the same core. The current algorithm for choosing cores (`takeByTopology()` function) tries to first allocate full cores and sockets based on the request size.